File: thesis/algorithms/typescript/np-completo_guloso.ts
import * as fs from 'fs';

interface Item {
    weight: number;
    value: number;
}

interface KnapsackData {
    capacity: number;
    items: Item[];
}

// Algoritmo guloso: ordena por valor/peso e preenche enquanto houver capacidade
function knapsack(items: Item[], capacity: number): number {
    const sorted = [...items].sort((a, b) => (b.value / b.weight) - (a.value / a.weight));

    let totalValue = 0;
    let currentWeight = 0;

    for (const item of sorted) {
        if (currentWeight + item.weight <= capacity) {
            currentWeight += item.weight;
            totalValue += item.value;
        }
    }

    return totalValue;
}

function main() {
    const size = process.argv[2] || 'small';
    const path = `datasets/${size}/knapsack.json`;

    if (!fs.existsSync(path)) {
        console.error('Arquivo não encontrado.');
        return;
    }

    const data: KnapsackData = JSON.parse(fs.readFileSync(path, 'utf-8'));
    const result = knapsack(data.items, data.capacity);
    console.log(`Valor aproximado (greedy) para ${data.items.length} itens (capacidade ${data.capacity}, ${size}): ${result}`);
}

main();

File: thesis/algorithms/python/np-completo_guloso.py
import json
import sys

def knapsack(items, capacity):
    # Ordenar por valor/peso decrescente
    items.sort(key=lambda x: x[1] / x[0], reverse=True)

    total_value = 0
    current_weight = 0

    for weight, value in items:
        if current_weight + weight <= capacity:
            current_weight += weight
            total_value += value

    return total_value

def main(size='small'):
    try:
        with open(f'datasets/{size}/knapsack.json', 'r') as f:
            data = json.load(f)
    except FileNotFoundError:
        print(f"Arquivo datasets/{size}/knapsack.json não encontrado.")
        return

    items = [(item["weight"], item["value"]) for item in data["items"]]
    capacity = data["capacity"]
    result = knapsack(items, capacity)
    print(f'Valor aproximado (greedy) para {len(items)} itens (capacidade {capacity}, {size}): {result}')

if __name__ == '__main__':
    size = sys.argv[1] if len(sys.argv) > 1 else 'small'
    main(size)
